web/src/hooks/viewport.js

```diff
@@ -1,49 +1,35 @@
-import { useState, useEffect } from 'react';
+import { useState, useEffect, useCallback } from 'react';
 
-export function useWidthObserver(ref, offset = 0) {
+export function useRefWidthHeightObserver(offsetX = 0, offsetY = 0) {
   const [width, setWidth] = useState(0);
+  const [height, setHeight] = useState(0);
+  const [node, setNode] = useState(null); // The DOM node
 
-  // Resize hook
-  useEffect(() => {
-    const updateWidth = () => {
-      if (ref.current) {
-        const newWidth = ref.current.getBoundingClientRect().width - offset;
-        if (newWidth !== width) { setWidth(newWidth); }
+  // See https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node
+  const ref = useCallback((newNode) => {
+    // This callback will be called once the ref
+    // returned has been attached to `node`.
+    const update = () => {
+      if (newNode !== null) {
+        const newWidth = newNode.getBoundingClientRect().width - offsetX;
+        const newHeight = newNode.getBoundingClientRect().height - offsetY;
+        if (width !== newWidth) { setWidth(newWidth); }
+        if (height !== newHeight) { setHeight(newHeight); }
+        if (node !== newNode) { setNode(newNode); }
       }
     };
     // Set initial width
-    updateWidth();
+    update();
     // Update container width on every resize
-    window.addEventListener('resize', updateWidth);
-    return () => {
-      window.removeEventListener('resize', updateWidth);
-    };
-  });
-
-  return width;
-}
-
-export function useHeightObserver(ref, offset = 0) {
-  const [height, setHeight] = useState(0);
-
-  // Resize hook
-  useEffect(() => {
-    const updateHeight = () => {
-      if (ref.current) {
-        const newHeight = ref.current.getBoundingClientRect().height - offset;
-        if (newHeight !== height) { setHeight(newHeight); }
-      }
-    };
-    // Set initial height
-    updateHeight();
-    // Update height on every resize
-    window.addEventListener('resize', updateHeight);
+    window.addEventListener('resize', update);
     return () => {
-      window.removeEventListener('resize', updateHeight);
+      window.removeEventListener('resize', update);
     };
-  });
+  }, [offsetX, offsetY]);
```

Comment:
I feel we should probably add `width`, `height` and `node` to the dependencies here as long as we're doing the inequality checks `if (var !== newVar)`, otherwise we might potentially run into inaccuracies comparing to the old values.

Alternatively, maybe all 3 checks could just be dropped to keep it simple as in the [docs example](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node)?

Reply:
Wouldn't that mean that `useCallback` is re-executed when the `width` changes, potentially returning a new `ref`?

Reply:
Yes, but the new ref wouldn't make any changes then, right? It's a bit clunky, I agree. So maybe it's just better to keep it simple and drop the `if` conditions :thinking: 

Just to illustrate why I think the current code is problematic, imagine we start with both `width` and `newWidth` set to 500 and then `newWidth` gets changed to 600, which triggers a call to `setWidth` which supposedly updates the `width` externally to 600 as well. But if my understanding is correct, since `width` is not in the dependencies, the cached callback function will not be changed and it will still be using the old value for `width` which is 500. So when `newWidth` gets changed back to 500, the if-condition will prevent `setWidth` to be called so `useRefWidthHeightObserver` will keep returning 600 instead of the correct width of 500 now. Does that make sense?

Reply:
@fbarl I think I understand. Are you saying that `width` inside the closure of `update = () { ... }` points to the *initial*  value of the `width` state, and thus, if the `width` state is not part of the dependencies, then the callback will still compare the initial `width`.
That seems sensible to me.

One thing to note though is that the returned `ref` is actually a callback function `(node) =>  { ... }` which gets executed once the `ref` (function) is attached to the DOM.
Changing the `ref` by instantiating a new `callback` will recreate a new callback that will never be called, as it will never be re-attached (DOM element is already created).

The simplest solution is to get rid of the `if` conditions.